Library/GeneralImage.cpp

```diff
@@ -296,6 +299,76 @@ bool GeneralImage::LoadImage(const std::wstring& imageName)
 	return false;
 }
 
+bool GeneralImage::LoadImageFromMeasure(Measure* m)
+{
+	if (!m_Skin || m == nullptr)
+	{
+		DisposeImage();
+		return false;
+	}
+
+	auto pm = dynamic_cast<MeasurePlugin*>(m);
+	if (pm == nullptr)
+	{
+		DisposeImage();
+		return false;
+	}
+
+	ImageOptions info;
+	info.m_Path = std::wstring{ L"measure://" } + m->GetSkin()->GetFilePath() + L"/" + m->GetName();
+	info.m_FileSize = 0;
+
+	// PluginImageData is a type to be created, must contain at least:
+	// 1. uint32_t pixels array
+	// 2. Width and height
+	// 3. uint64_t timestamp
+	PluginImageData im{};
+	bool success = pm.GetImage(&im);

```

Comment:
class "MeasurePlugin"  has no member `GetImage`
What is this function going to do (as in functionality)?
Also should the plugin provide the definition? or i should write that in `MeasurePlugin` class?

Reply:
`MeasurePlugin` is a wrapper over plugins.
To be honest, yesterday I didn't look into how it works, but few years ago this class was just holding pointers to plugin functions, and using these pointers with some work to make sure that plugin gets a valid data.

Reply:
>Also should the plugin provide the definition?

Plugin should provide some sort of `GetImage` exported function.
Class `MeasurePlugin` will find this function and use it, when its `GetImage` method is called.
I guess it's kinda confusing that I don't know how to name these things differently, I hope you understand what I mean.

Reply:
>I guess it's kinda confusing that I don't know how to name these things differently, I hope you understand what I mean.

Yep i guess i got it:

`MeasurePlugin` class has a _virtual_ void method called `GetImage`, when the class calls it, it fills `PluginImageData` struct with the necessary data, then `LoadImageFromMeasure` will read this data and generate an image.

But there is something i still don't get, if that function is actually virtual does it mean the plugin "must" provide an impl for it?

Reply:
`MeasurePlugin` has an plugin library handle, and it can search for plugin functions using this handle.
It writes something like `getFunction(dllHandle, "GetImage")`, saves the function pointer into its field and then when its method `GetImage` is called it uses this function pointer.
Look at how other plugin functions are handled inside `MeasurePlugin` class.

Reply:
My LocalPluginLoader plugin works the same way.

Reply:
>My LocalPluginLoader plugin works the same way.

I thought you were using some sort of black magic lol

Reply:
>Look at how other plugin functions are handled inside `MeasurePlugin` class.

Sorry for asking much but i didn't find any reference to learn from.
My main question is, how am i supposed to create `GetImage` function?

Reply:
Finding plugin function in dll: https://github.com/LI7XI/rainmeter/blob/cd9d25d3d91e3b96a7325efbbf8af0eef49f81ea/Library/MeasurePlugin.cpp#L132-L135

Probably, the only thing `GetImage` will do is call the plugin function with appropriate args. Like, literally.
Somewhat like this:
```cpp
bool MeasurePlugin::GetImage(ImageData& im)
{
	if (m_GetImageFunc)
	{
		return m_GetImageFunc(m_PluginData, &im) != nullptr;
	}

	return false;
}
```
And now the question is how you will design the interface of the function.

Reply:
Maybe a cheap and safe way would be to define function like this:
```cpp
struct PluginImageData {
WCHAR zero;
// image data goes here
};
LPCWSTR GetImage(void* pluginData, int one, PluginImageData* im)
```
This way the function will be compatible with current convention for custom functions.
`int one` will always be 1.
PluginImageData.zero will always contain '\0' symbol. This way functions that expect `LPCWSTR(*CUSTOMFUNCTION)(void*, const int, const WCHAR* argv[])` (definition from plugin measure header) will work correctly. `PluginImageData* im` will contain zero character at the beginning, so they will treat it as a string of length 0.